### ISSUE TITLE
Pattern Miner shallow clean-ups

### ISSUE THREAD

**Issue by ngeiswei** [opencog/opencog#2787]
I noticed the following possible clean-ups in the pattern miner. They
are shallow but should already help to increase readability.

- [x] In `PatternMiner::cleanUpPatternMiner()` a lot of swap is taking
  place, why is that? Is it because it is faster than clear()? Is it
  to deallocate memory as explained in
  https://stackoverflow.com/questions/7351143/whats-faster-recreate-or-clear#7351163
  ? If there is an advantage wrap these swaps into a `swap_clear()`
  method, or add a generic version in cogutils.
- [ ] Strive to have 80 col max. Otherwise it wastes a lot of screen
  estate, you can't put 2 pages side-by-side, editor auto-wrapping
  just doesn't cut it. Or suffice to say it's in the development
  standard http://wiki.opencog.org/w/Development_standards#Formatting
- [x] Replace `map<Handle, Handle>` by `HandleMap`, `pair<Handle,
  Handle>` by `HandlePair`, etc, as defined in `Handle.h`
- [x] Use by C++11 range loop when possible, such as "for (iter =" in
  `PatternMinerDF.cc`. Look also at `niter`, `ValueToVarIt`,
  `oneSuperRelationIt`, etc.
- [x] Simplify insert pair inside a `std::map` (no need to rebuild an
  unmodified pair, like in l.226 of `PatternMinerBF.cc`).
- [x] Replace `0` and `NULL` by `nullptr` (look for `= 0` and `NULL`)
- [x] Replace `node_types_should_not_be_vars` vector by set
- [x] `incomeingPtr` -> `incomingPtr`
- [x] simplify var_num initialization using ternary conditional
  operator (? :)
- [x] Simplify template closing brackets `> >` -> `>>`
- [x] use `boost::sort` instead of `std::sort`. This is optional, some
  people like Linas don't like that we use boost when we could use the
  standard, because boost isn't as stable, but this one is pretty
  stable, and will likely end-up in the standard anyway.
- [x] In `initPatternMiner` use resize instead of pushing back
  `patternVector`
- [ ] Constify the whole thing
  Like loops in `PatternMiner::calculateSurprisingness`
- [ ] Define a PatternMiner logger and use it instead of printing on
  the stdout.
- [x] Simplify `PatternMiner::containWhiteKeywords` using
  `PatternMiner::containKeywords`.
- [x] Factorize
    ```
    node_types_str.erase(std::remove(node_types_str.begin(), node_types_str.end(), ' '), node_types_str.end());
    vector<string> typeStrs;
    boost::split(typeStrs, node_types_str, boost::is_any_of(","));
    ```
- [x] Reimplement `PatternMiner::isInHandleSeq`,
  `PatternMiner::isIgnoredType`, `PatternMiner::isTypeInList`,
  `PatternMiner::isIgnoredContent` with `is_in` from `cogutils`.
- [x] Replace the range loop over `allNodesInEachLink` in
  `PatternMiner::splitDisconnectedLinksIntoConnectedGroups` by
  `is_disjoint` from `cogutils`.
- [ ] Better use sets rather than sequences in
  `PatternMiner::splitDisconnectedLinksIntoConnectedGroups`
- [x] `PatternMiner::splitDisconnectedLinksIntoConnectedGroups` could
  be renamed into `PatternMiner::partitionIntoCliques`
- [x] Have `linktype_black_list` and `linktype_white_list` be
  `std::set` instead if `std::vector` and Reimplement
  `PatternMiner::add_linktype_to_white_list` with
  `std::pair<iterator,bool> std::set::insert(const value_type& value)`
  and `PatternMiner::remove_Ignore_Link_Type` with `std::set::erase`
- [x] Do similar simplifications with
  `PatternMiner::add_link_type_to_same_link_types_not_share_second_outgoing`,
  `PatternMiner::remove_link_type_from_same_link_types_not_share_second_outgoing`,
  `PatternMiner::add_node_type_to_node_types_should_not_be_vars`,
  `PatternMiner::remove_node_type_from_node_types_should_not_be_vars`,
  `PatternMiner::add_keyword_to_black_list`,
  `PatternMiner::remove_keyword_from_black_list`,
  `PatternMiner::add_keyword_to_white_list`,
  `PatternMiner::remove_keyword_from_white_list`
- [ ] Move code like
  ```
  newHTreeNode->parentLinks.insert(parentNode);
  parentNode->childLinks.insert(newHTreeNode);
  ```
  in a method of `HTreeNode` or `HTree`
- [x] Replace
        ```
        float abs_diff = diff;
        if (abs_diff < 0)
            abs_diff = - abs_diff;
        ```
  by `std::abs`
- [x] Remove `find` calls before inserting to an ordered container. No need to check if an element is already in the container as the `insert` will do the check. Removing these checks results in simpler and faster code.

There are other less shallow changes but they require more thinking (by me anyway).

- [ ] Look into `XXX why do we need to set the TV ???` or
  `TruthValue::TRUE_TV()`
- [ ] Why representing pattern key as string instead of atom?
- [ ] Likewise why use a string as key of `allEntityNumMap` instead of
  atom?
- [ ] Decouple running the algorithm and outputting its
  results. For instance move out calls to
  `PatternMiner::OutPutFrequentPatternsToFile` within
  `PatternMiner::runPatternMiner` or such.

@shujingke I don't necessarily suggest that you take care of them. In fact I would perhaps prefer to do it because it would help me to get familiar with the code. But, first, I need your feedback on a couple of stuff (like the swap thing), and second, I really won't do it unless the utest passes, and I need your help on that. Also of course it would be better if you don't have major changes not merged yet. What do you think?

**Comment by linas** [comment]
Also, in #2800 it appears this code mis-uses PATTERN_LINK which is defined as a kind of SCOPE_LINK.  This mis-use might be hiding bugs, or might be making the code run a lot more slowly than it could otherwise. (because PATTERN_LINK has a complex constructor)

**Comment by linas** [comment]
Also, can't tell if its in the list above, most of the usage of the config file should be removed.  pretty much all uses of strings in the API, e.g. black_list, white_list, etc. should be replaced by atoms of some kind.

**Comment by shujingke** [comment]
In PatternMiner::cleanUpPatternMiner() a lot of swap is taking
place, it's because usually those contains are huge, and basd on my experiences before sometimes clean() do not really make the system reclaim the memory. But maybe C++ 11 already fix this..

**Comment by shujingke** [comment]
If boost:sort will likely end-up in the standard anyway, then why shall we use it when we can just use std::sort ?

**Comment by shujingke** [comment]
XXX why do we need to set the TV ??? or
TruthValue::TRUE_TV() 
I don't know. This was someone else changed my code. I didn't have this at the beginning.

**Comment by shujingke** [comment]
Linas, "Also, most of the usage of the config file should be removed. Pretty much all uses of strings in the API, e.g. black_list, white_list, etc. should be replaced by atoms of some kind."  I don't get it. The value of  black_list, white_list are atoms of some kind.

**Comment by ngeiswei** [comment]
+ Regarding swap: I don't think C++11 takes care of that. So this should be wrapped in a clean_by_swap method, template or something.
+ `boost::sort` and other boost range algorithms are simpler to call than their `std` counterparts. For example `boost::sort(my_container)` instead of `boost::sort(my_container.begin(), my_container.end())`, but really, as I said, this is no big deal, they both have pros and cons, std is still more portable while boost is more compact.
+ I seem to agree with Linas about the black_list and white_list, it seems it would be more elegant if these were atoms, but I think we can leave that aside for now. A lot of simplification can take place before dealing with that.

@shujingke , would you rather take care of the proposed changes by yourself? Or let them to me? Or take care of some and let me some others? I'm fine any ways, and I'll be available to work on that probably next week (I still have my head into the inference control learning example I'm writing to provide you a full corpus of inferences, which will be substantially different than what I initially sent you BTW).

**Comment by linas** [comment]
@shujingke  Currently, everything in the `PatternMinerSCM.cc` uses strings for the API.  It should use atoms, not strings. So, for example, for `keyword_white_list` instead of writing "office,president,chairman"  you could write:
```
ListLink
    Concept "office"
    Concept "president"
    Concept "chairman"
```
It doesn't have to be a ListLink, it could be a SetLink, I don't care. It doesn't have to be a ConceptNode, it could be a PredicateNode or a WordNode, I don't care.  What I do care is that it should not be a string.

**Comment by linas** [comment]
@shujingke regarding the config file: you should try to not use the config file for anything.  There are multiple problems with the config file:

* It is read only once, when the cogserver is started. Thus, it is impossible to change the values in it after the cogserver is started. You would have to stop the cogserver, and restart it.

* The config file can only be used with the cogserver. If you don't use the cogserver, then the config file doesn't work. Many people do not use the cogserver, at all. The python module does not even allow the cogserver to be started!

* The config file has use-limitations. You asked "Is there a way to direct assign an atom in the config file?" and the answer is no. I do not want to enhance the config file to add this feature. Or anything else that people have asked for: assign TV's, or to "pre-run"  commands, or anything else. If one starts adding these new abilities to the config file, then one is essentially designing a new, and bad, programming language. I don't want to do that.

So, for example: if you store `keyword_white_list` as atoms, then you can use either scheme or python instead of the config file.  And you can change this list at any time, without having to restart the cogserver. And finally, you would not have to even run the cogserver.

From what I can tell, the pattern miner never actually uses the cogserver for anything at all, so why are you tied to it?  Why would you want to be tied to it?

**Comment by ngeiswei** [comment]
I mostly agree with @linas . The only detail of rectification is that the config API does allow to dynamically change the config values, so one can change the values once the cogserver has started. However this only true from within C++ programs, not scheme since the config has no binding for it, and one might indeed wonder the point of such bindings given that scheme can be used as a much more powerful configuration language.

I totally agree about removing the prerequisite of having to run the cogserver to use the pattern miner.

**Comment by ngeiswei** [comment]
I'm closing as this will probably never be completed.
